command/operator_api.go

```diff
@@ -0,0 +1,446 @@
+    Output curl command to stdout and exit.
+    HTTP Basic Auth will never be output. If the $NOMAD_HTTP_AUTH environment
+    variable is set, it will be referenced in the appropriate curl flag in the
+    output.
+    ACL tokens set via the $NOMAD_TOKEN environment variable will only be
+    referenced by environment variable as with HTTP Basic Auth above. However
+    if the -token flag is explicitly used, the token will also be included in
+    the output.
+
+  -filter <query>
+    Specifies an expression used to filter query results.
+
+  -H <Header>
+    Adds an additional HTTP header to the request. May be specified more than
+    once. These headers take precedent over automatically ones such as
+    X-Nomad-Token.
+
+  -verbose
+    Output extra information to stderr similar to curl's --verbose flag.
+
+  -X <HTTP Method>
+    HTTP method of request. Defaults to GET.
+`
+
+	return strings.TrimSpace(helpText)
+}
+
+func (*OperatorAPICommand) Synopsis() string {
+	return "Query Nomad's HTTP API like curl"
+}
+
+func (c *OperatorAPICommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-dryrun": complete.PredictNothing,
+		})
+}
+
+func (c *OperatorAPICommand) AutocompleteArgs() complete.Predictor {
+	//TODO(schmichael) wouldn't it be cool to build path autocompletion off
+	//                 of our http mux?
+	return complete.PredictNothing
+}
+
+func (*OperatorAPICommand) Name() string { return "operator api" }
+
+func (c *OperatorAPICommand) Run(args []string) int {
+	var dryrun bool
+	var filter string
+	headerFlags := newHeaderFlags()
+
+	flags := c.Meta.FlagSet(c.Name(), FlagSetClient)
+	flags.Usage = func() { c.Ui.Output(c.Help()) }
+	flags.BoolVar(&dryrun, "dryrun", false, "")
+	flags.StringVar(&filter, "filter", "", "")
+	flags.BoolVar(&c.verboseFlag, "verbose", false, "")
+	flags.StringVar(&c.method, "X", "GET", "")
+	flags.Var(headerFlags, "H", "")
+
+	if err := flags.Parse(args); err != nil {
+		c.Ui.Error(fmt.Sprintf("Error parsing flags: %v", err))
+		return 1
+	}
+	args = flags.Args()
+
+	if len(args) < 1 {
+		c.Ui.Error("A path or URL is required")
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	if n := len(args); n > 1 {
+		c.Ui.Error(fmt.Sprintf("operator api accepts exactly 1 argument, but %d arguments were found", n))
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	// By default verbose func is a noop
+	verbose := func(string, ...interface{}) {}
+	if c.verboseFlag {
+		verbose = func(format string, a ...interface{}) {
+			// Use Warn instead of Info because Info goes to stdout
+			c.Ui.Warn(fmt.Sprintf(format, a...))
+		}
+	}
+
+	// Opportunistically read from stdin and POST unless method has been
+	// explicitly set.
+	stat, _ := os.Stdin.Stat()
+	if (stat.Mode() & os.ModeCharDevice) == 0 {
+		verbose("* Reading request body from stdin.")
+		c.body = os.Stdin
+		if c.method == "" {
+			c.method = "POST"
+		}
```

Comment:
This is harmless, but `c.method` is defaulted to `GET` in the flag parser, right?

Reply:
Yes, `-X` is default to `GET` on line 109 above. This defaults it to POST if there's data on stdin... curl doesn't do that trick though, so I'm not sure if this should?

Reply:
Maybe I'm misunderstanding, but I'm saying line 145 here `c.method == ""` will never be true, because of line 109. To get what you want, the condition would need to be `c.method == "GET"`.

Reply:
Yeah, I think @shoenig is right. `GET` is the flag default, so it will never be `""` here. 

```console
$ cat ns.json | nomad operator api /v1/namespace
Invalid method% 
```

Reply:
Ha, thanks! Will fix.